benchmark/microbenchmark/src/androidTest/java/com/apollographql/apollo3/benchmark/ApolloStoreIncubatingTests.kt

```diff
@@ -0,0 +1,98 @@
+package com.apollographql.apollo3.benchmark
+
+import androidx.benchmark.junit4.BenchmarkRule
+import androidx.benchmark.junit4.measureRepeated
+import androidx.test.platform.app.InstrumentationRegistry
+import com.apollographql.apollo3.api.json.jsonReader
+import com.apollographql.apollo3.api.parseJsonResponse
+import com.apollographql.apollo3.benchmark.Utils.dbName
+import com.apollographql.apollo3.benchmark.Utils.operationBasedQuery
+import com.apollographql.apollo3.benchmark.Utils.resource
+import com.apollographql.apollo3.benchmark.test.R
+import com.apollographql.apollo3.cache.normalized.incubating.ApolloStore
+import com.apollographql.apollo3.cache.normalized.incubating.api.CacheKeyGenerator
+import com.apollographql.apollo3.cache.normalized.incubating.api.CacheResolver
+import com.apollographql.apollo3.cache.normalized.incubating.api.FieldPolicyCacheResolver
+import com.apollographql.apollo3.cache.normalized.incubating.api.MemoryCacheFactory
+import com.apollographql.apollo3.cache.normalized.incubating.api.NormalizedCacheFactory
+import com.apollographql.apollo3.cache.normalized.incubating.api.TypePolicyCacheKeyGenerator
+import com.apollographql.apollo3.cache.normalized.incubating.sql.SqlNormalizedCacheFactory
+import org.junit.Assert
+import org.junit.Rule
+import org.junit.Test
+import java.lang.reflect.Method
+import java.util.concurrent.Executors
+
+class ApolloStoreIncubatingTests {
+  @get:Rule
+  val benchmarkRule = BenchmarkRule()
+
+  @Test
+  fun concurrentReadWritesMemory() {
+    concurrentReadWrites(MemoryCacheFactory())
+  }
+
+  @Test
+  fun concurrentReadWritesSql() {
+    Utils.dbFile.delete()
+    // Pass context explicitly here because androidx.startup fails due to relocation
+    val cacheFactory = SqlNormalizedCacheFactory(InstrumentationRegistry.getInstrumentation().context, dbName)
+    concurrentReadWrites(cacheFactory)
+  }
+
+  @Test
+  fun concurrentReadWritesMemoryThenSql() {
+    Utils.dbFile.delete()
+    val cacheFactory = MemoryCacheFactory().chain(SqlNormalizedCacheFactory(InstrumentationRegistry.getInstrumentation().context, dbName))
+    concurrentReadWrites(cacheFactory)
+  }
+
+  private fun concurrentReadWrites(cacheFactory: NormalizedCacheFactory) {
+    val apolloStore = createApolloStore(cacheFactory)
+    val query = operationBasedQuery
+    val data = query.parseJsonResponse(resource(R.raw.calendar_response_simple).jsonReader()).data!!
+    val threadPool = Executors.newFixedThreadPool(CONCURRENCY)
+    benchmarkRule.measureRepeated {
+      val futures = (1..CONCURRENCY).map {
+        threadPool.submit {
+          // Let each thread execute a few writes/reads
+          repeat(WORK_LOAD) {
+            apolloStore.writeOperation(query, data)
+            val data2 = apolloStore.readOperation(query)
+            Assert.assertEquals(data, data2)
+          }
```

Comment:
Can you also add an integration test, that tests end to end using `apolloClient.query()` and the default dispatcher? (should be `Dispatchers.Default` IIRC). Are the results similar?

Reply:
Do you mean still in a microbenchmark or something else?

Reply:
I think so? Same thing but closer to the real scenario of executing a query. Probably using MockWebServer or so. I know it adds more variance but it's also what users are actually doing.

Reply:
Added in c8cee0f74690b5fd78f3bd9b4bd76f2884f37548 benches that execute queries in parallel. The results are a bit questionable:

|             | Memory     | Sql           | Memory then sql |
|-------------|------------|---------------|-----------------|
| Before      | 44,201,994 | 1,047,540,153 | 41,335,067      |
| After       | 40,916,743 | 1,079,325,820 | 38,458,753      |
| Improvement | 7.4%       | -3%           | 6.9%            |